src/main/java/io/jenkins/plugins/pipeline_elasticsearch_logs/write/fluentd/FluentdWriter.java

```diff
@@ -136,87 +154,94 @@ public String getDisplayName() {
     }
 
     /**
-     * Logs the given object to Fluentd. This method blocks if Fluentd is not reachable and
-     * throws an IOException after TIMEOUT_MILLIS.
+     * Logs the given object to Fluentd asynchronously
+     * Checks if there have been any data that couldn't be sent after several retries and then fails.
      *
      * @param data The data to post
      * @throws IOException if something failed permanently
      */
     @Override
     public void push(Map<String, Object> data) throws IOException {
-        if(fluentd == null || configChanged) initFluentdLogger();
-        try {
-
-            long timeMarker = System.currentTimeMillis();
-            boolean sent = fluentd.log(tag, data);
-            if(!sent) {
-                //Errors are not reported explicitly. But the latest error should be the reason for (sent == false)
-                Exception lastError = errorHandler.getLastError(timeMarker);
-                throw new IOException("Could not send logs to fluentd", lastError);
+        if(fluentd == null) initFluentdLogger();
+        int count = 0;
+        while (true) {
+            try {
+                fluentd.emit(tag, data);
+                count++;
+                if (count > maxRetries) {
+                    throw new IOException("Not able to emit data after " + maxRetries + " tries.");
+                }
+                break;
             }
+            catch (BufferFullException e) {
```

Comment:
We only catch `BufferFullException`. Is this the only one we should catch?

Btw. `BufferFullException` extends `IOException`. Whats the meaning of the `RetryableException`? Documentation there is quite bad. And I did not see cases where a `RetryableException` is thrown in their code which would affect us. Therefore I also don't get the `checkForRetryableException()` below...

Reply:
I think yes we should only catch BufferFullException. This is thrown when the event couldn't be written to the internal buffer (maybe because it was too big, though we try to avoid this upfront).
The RetryableException is thrown when the library was not able to flush and send 30 times in a row. It's a bit hidden in the sender classes.

Reply:
So the name does not mean, it is worth to retry, when such an Exception is caught. But it means the retry already happened several times and did not resolve the problem?